### ISSUE TITLE
Reproducibility with DGL implementation

### ISSUE THREAD

**Issue by WMX567** [dmlc/dgl#5425]
DGL: 0.6.1
PyTorch: 1.12.1

I used the GCN Predictor implemented with DGL in this [repo](https://github.com/QueuQ/CGLB/tree/master/GCGL/Backbones).

I set my random seed:

    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed+r)
    torch.cuda.manual_seed_all(seed+r)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
    dgl.random.seed(seed+r) 

However, I could not reproduce my results. Every time, the accuracies are different. Any idea about how to reproduce the results?

**Comment by BarclayII** [comment]
Could you try upgrading to the latest stable version (1.0.1)?

**Comment by WMX567** [comment]
How to upgrade? Thanks! Which command should I use?

**Comment by BarclayII** [comment]
You could upgrade it with
```
conda uninstall dgl-cudaXX.X  # your CUDA version
conda install dgl -c dglteam/label/cuXXX # your CUDA version
```

**Comment by WMX567** [comment]
I upgraded to 1.0.1 but the problems are still there. I could not reproduce the results. I change the fn.copy_src to fn.copy_u. I am not sure whether these changes will be problems.

**Comment by BarclayII** [comment]
I only saw the model definition there.  The code seems to be based on dgl-lifesci.  How did you run the model?  Maybe it's because you have used DataLoader with multiprocessing, which has some inherent randomness.

**Comment by frozenbugs** [comment]
Hi @WMX567 , is the problem resolved? Let us know if you have more questions.

**Comment by xjtubraveheart** [comment]
I encountered the same problem. When the GNN-related code is commented out, the results, whether GPU or CPU, can be reproduced exactly in each case. Once the GNN learning code is added, the results with GPU start to fluctuate and cannot be reproduced. However, the results with CPU can be reproduced. DGL version: 1.0.1+cu113, Torch version: 1.11.0+cu113.

**Comment by BarclayII** [comment]
@xjtubraveheart Could you post your code?  And could you also upgrade DGL to the latest version (1.0.2)?

**Comment by xjtubraveheart** [comment]
def setup_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
    dgl.seed(seed)

class GeoGCNLayer(nn.Module):
    def __init__(self,
                g,
                args,
                device='cuda'
                ):
        super(GeoGCNLayer, self).__init__()
        self.g = g
        self.device = device
        self.act = nn.LeakyReLU(0.2)
        self.is_att = args.is_att #地理空间影响和时序转移影响融合是否采用注意力方式
        self.is_sgc = args.is_sgc #是否采用SGC方式，即去掉非线性变化处理
        self.args = args
        self.is_lightgcn = args.is_lightgcn
        if self.is_att:
            self.attn_fuse = SemanticAttention(args.loc_dim, args.loc_dim*4)
        
    def forward(self, feat):
        funcs = {}
        feat_t = feat if self.is_lightgcn else self.feat_tran(feat)
        self.g.ndata['f'] = feat_t
        for srctype, etype, dsttype in self.g.canonical_etypes:
            if etype == 'geo':
                funcs[etype] = (fn.copy_u('f', 'm'), fn.mean('m', 'geo'))
            elif etype == 'cat':
                funcs[etype] = (fn.copy_u('f', 'm'), fn.mean('m', 'cat'))
            else:
                funcs[etype] = (fn.u_mul_e('f', 'w', 'm'), fn.sum('m', 'trans'))
                    
        self.g.multi_update_all(funcs, 'sum')
        if self.is_att: 
            geo = self.g.ndata['geo'].unsqueeze(1)
            trans = self.g.ndata['trans'].unsqueeze(1)
            if not self.args.base and self.args.cp4:
                cat = self.g.ndata['cat'].unsqueeze(1)
                z = torch.cat([geo, trans, cat], 1)
            else:
                z = torch.cat([geo, trans], 1)
            feat = self.attn_fuse(z)
        return feat if self.is_sgc else self.act(feat)

class SemanticAttention(nn.Module):
    def __init__(self, in_size, hidden_size=128):
        super(SemanticAttention, self).__init__()

        self.project = nn.Sequential(
            nn.Linear(in_size, hidden_size),
            nn.Tanh(),
            nn.Linear(hidden_size, 1, bias=False)
        )

    def forward(self, z):
        w = self.project(z).mean(0)                    
        beta = torch.softmax(w, dim=0)                 
        beta = beta.expand((z.shape[0],) + beta.shape) 
        return (beta * z).sum(1)

class GeoGCN(nn.Module):
    def __init__(self,
                g,
                tran_e_w,
                args,
                device='cuda'
                ):
        super(GeoGCN, self).__init__()
        g = g.int()
        g = dgl.remove_self_loop(g, etype='geo')
        g = dgl.add_self_loop(g, etype='geo')
        self.g = g.to(device)
        self.g.edges['trans'].data['w'] = torch.tensor(tran_e_w).float().to(device)
        self.num_layer = args.GeoGCN_layer_num
        self.device = device
        self.act = nn.LeakyReLU(0.2)
            
        self.gcn = nn.ModuleList()
        for i in range(self.num_layer):
            self.gcn.append(
            GeoGCNLayer(self.g, args, device)
        )
            
    def forward(self, feat):
        for i in range(self.num_layer - 1):
            feat = self.gcn[i](feat)
        feat = self.gcn[-1](feat)
        return feat

loc_emb_model = nn.Embedding(data.loc_num, args.loc_dim).to(device)
geogcn_model = GeoGCN(data.loc_g, data.tran_edge_weight, args, device).to(device)
loc_emb_model.train()
geogcn_model.train()
loc_emb = loc_emb_model(torch.tensor(range(data.loc_num)).to(device))
loc_emb = geogcn_model(loc_emb)

loc_emb_model.eval()
geogcn_model.eval()
loc_emb = loc_emb_model(torch.tensor(range(data.loc_num)).to(device))
loc_emb = geogcn_model(loc_emb)

The approximate code flow is as shown above. Next,  I try to update DGL to the latest version (1.0.2) and retest the experimental results. @BarclayII

**Comment by BarclayII** [comment]
I think our GPU kernel is implemented non-deterministically and due to floating point errors the results naturally fluctuate depending on the order of arithmetics.  How much did the resulting metric fluctuate?  Normally this shouldn't be an issue.

**Comment by luolo-cry** [comment]
I hava the same problem.

**Comment by WMX567** [comment]
> Hi @WMX567 , is the problem resolved? Let us know if you have more questions.

No, the problem is still there. I could not reproduce it.

**Comment by luolo-cry** [comment]
is this problem fixed？

**Comment by BarclayII** [comment]
> is this problem fixed？

Our GPU kernel is implemented non-deterministically so each time the result might be different.  That being said, how much did the resulting metric fluctuate?  If it changes a lot then this is a problem we should look into.  @luolo-cry

**Comment by jermainewang** [comment]
I am closing this issue assuming you are happy about our response. Feel free to follow up and reopen the issue if you have more questions with regard to our response.
